Library/Source/ShaderCompilerOpenGL.cpp

```diff
@@ -1,19 +1,105 @@
 #include "ShaderCompiler.h"
+#include "ResourceLimits.h"
+#include <arcana/experimental/array.h>
+#include <glslang/Public/ShaderLang.h>
+#include <SPIRV/GlslangToSpv.h>
+#include <spirv_parser.hpp>
+#include <spirv_glsl.hpp>
+#include "Console.h"
 
 namespace babylon
 {
+    extern const TBuiltInResource DefaultTBuiltInResource;
+
+    namespace
+    {
+        void AddShader(glslang::TProgram& program, glslang::TShader& shader, std::string_view source)
+        {
+            const std::array<const char*, 1> sources{ source.data() };
+            shader.setStrings(sources.data(), gsl::narrow_cast<int>(sources.size()));
+            shader.setEnvInput(glslang::EShSourceGlsl, shader.getStage(), glslang::EShClientOpenGL, 100);
+            shader.setEnvClient(glslang::EShClientOpenGL, glslang::EShTargetOpenGL_450);
+            shader.setEnvTarget(glslang::EShTargetSpv, glslang::EShTargetSpv_1_0);
+
+            if (!shader.parse(&DefaultTBuiltInResource, 430, false, EShMsgDefault))
+            {
+                Console::Error(shader.getInfoDebugLog());
+                throw std::exception();
+            }
+
+            program.addShader(&shader);
+        }
+
+        std::unique_ptr<spirv_cross::Compiler> CompileShader(glslang::TProgram& program, EShLanguage stage, std::string& glsl)
+        {
+            std::vector<uint32_t> spirv;
+            glslang::GlslangToSpv(*program.getIntermediate(stage), spirv);
+
+            spirv_cross::Parser parser{ std::move(spirv) };
+            parser.parse();
+
+            auto compiler = std::make_unique<spirv_cross::CompilerGLSL>(parser.get_parsed_ir());
+
+            // remove "uniform Frame { .... }" . Keep the uniforms
+            const std::string frame = "uniform Frame {";
+            const std::string frameEnd = "};";
+            size_t pos = glsl.find(frame);
+            if (pos != std::string::npos)
+            {
+                glsl.replace(pos, frame.size(), "");
+                pos = glsl.find(frameEnd, pos);
+                if (pos != std::string::npos)
+                {
+                    glsl.replace(pos, frameEnd.size(), "");
+                }
+            }
```

Comment:
The `uniform Frame` thing is added on the JavaScript side. It seems like maybe we should not do some of the processing on the JavaScript side instead of undoing it here for OpenGL. Or maybe we should not target OpenGL at all and only support Vulkan.

Reply:
I need the structure for Metal and Vulkan (maybe d3d also). Removing OpenGL is not a good idea as there are still many Android devices that do not support Vulkan properly (or at all). See https://gist.github.com/zeux/7d3ff56568c6b02de3a87cc69899d949

Reply:
Makes sense. I still think this processing should happen on the JS side. We can pass something to the JS side to tell it what to do with the shader processing.

Reply:
I agree. We can provide a JS function that returns the rendering API and do string/regex work based on that.